Add limit for returning the first 10 access points with high rssi

fabik111 opened the pull request:
During the `WiFi.scanNetworks()` the access points found are stored on a `vector `container. 

The allocation of memory for the `vector` doesn't grow linearly with the increasing of data stored, but logarithmically (view [this ](https://cplusplus.com/reference/vector/vector/) ) and this can hang up the Arduino UNO R4 board for memory saturation when performing a scan inside an enviroment with many APs.

This PR limits the number of element stored inside the vector for returning only the APs with the highest RSSI.

maidnl reviewed (changes requested):
@fabik111 just a little clarification about growth: anything that grows logarithmic-ally 
grows **less** than linear.
That is impossible for any container: for 1 element (big as "space") the vector needs at least 1 space, for 2 elements at least 2 spaces, for 3 elements at least 3 spaces and
so on (this is a linear growth).
If the growth would be logarithmic-ally that would mean that for 100 elements less 
than 100 spaces were required (which is obviously impossible).
What you mentioned in your PR said that "reallocations should only happen at 
logarithmic-ally growing intervals of size" which has a completely different meaning 
(although it is related to the fact that to store n elements more than n-space can be allocated, so the growth is more than linear and definitely more than logarithmic).

Your concern about saving memory is really appreciable, but is always 
risky: what if you are in a environment with a lot of (good) Wifi Network available 
and you have to connect with the 13-th RSSI strongest network? You cannot connect to the network. Did you run out memory due to the presence of so many Wifi Network? 
Or is just a concern? That is something to think about with this kind of PR.

Last but most important your code does not work: my understanding is that you want  to select the 12 strongest Wifi Networks, but your code does not achieve this result. I have tried it.
Here is the output of your algorithm (I limited the number of network to 4) 
with some made up values:
```
adding AP with RSSI 384
adding AP with RSSI 887
adding AP with RSSI 778
adding AP with RSSI 916
adding AP with RSSI 2004
adding AP with RSSI 2005
adding AP with RSSI 2006
adding AP with RSSI 2007
adding AP with RSSI 2008
adding AP with RSSI 2009

ACCESS POINTS Selected: 
384
887
778
2009
```
Clearly not the best 4 RSSI are selected.


Pay attention to the fact that
```
std::vector<CAccessPoint>::reverse_iterator rit = access_points.rbegin();
access_points.insert(rit.base(), ap);
```
is equivalent to 
```
access_points.push_back(ap);
```
just with a more convolute syntax. 
You can verify that with something like:
```
vector<int> v;
vector<int> w;

for(int i = 0; i < 10 ; i++){
  auto rit = v.rbegin();
  v.insert(rit.base(), i);

  w.push_back(i);
}

for(int i = 0; i < 10 ; i++){
  cout << v[i] << " " << w[i] << endl;
}
```
At the end the 2 vectors (w,v) are equal.
So you are always working on the last elements of the vector, but this does not allow you to select the strongest n Wifi Networks.

My personal suggestion, if really you run out of memory, is to completely revise the strategy: perhaps the access_points vector might be erased when is no more needed 
(perhaps when the board is connected to a Wifi network, but this is something that needs to be investigated in deep to see if there are not side effects).

Thank you

fabik111 commented:
Hello @maidnl, thank you for your review. 
I'm sorry I didn't express myself well in explaining the problem and I thank you for pointing out the error.

About the problem: this is not a concern, but actually what happens. 
In my apartment (in a condominium) I run out of memory during a scan (the board freezes), more precisely, while inserting the APs in the vector.
Here the log:
```
15:17:10.442 -> Scanning
// All the print comes from the CWifi::scanNetworks() method
15:17:10.487 -> CWifi::scanNetworks() - before modem write
15:17:10.487 -> - SRAM left: 12115
15:17:17.177 -> CWifi::scanNetworks() - after modem write
15:17:17.177 -> - SRAM left: 8223
15:17:17.177 -> discovered APs:18
15:17:17.177 -> inserting AP:0
15:17:17.177 -> - SRAM left: 7247
15:17:17.177 -> inserted AP
15:17:17.177 -> - SRAM left: 7247
15:17:17.316 -> inserting AP:1
15:17:17.316 -> - SRAM left: 7247
15:17:17.316 -> inserted AP
15:17:17.316 -> - SRAM left: 6983
15:17:17.316 -> inserting AP:2
15:17:17.316 -> - SRAM left: 6983
15:17:17.316 -> inserted AP
15:17:17.441 -> - SRAM left: 6463
15:17:17.441 -> inserting AP:3
15:17:17.441 -> - SRAM left: 6463
15:17:17.441 -> inserted AP
15:17:17.441 -> - SRAM left: 6463
15:17:17.441 -> inserting AP:4
15:17:17.441 -> - SRAM left: 6463
15:17:17.565 -> inserted AP
15:17:17.565 -> - SRAM left: 5431
15:17:17.565 -> inserting AP:5
15:17:17.565 -> - SRAM left: 5431
15:17:17.565 -> inserted AP
15:17:17.565 -> - SRAM left: 5431
15:17:17.565 -> inserting AP:6
15:17:17.704 -> - SRAM left: 5431
15:17:17.704 -> inserted AP
15:17:17.704 -> - SRAM left: 5431
15:17:17.704 -> inserting AP:7
15:17:17.704 -> - SRAM left: 5431
15:17:17.704 -> inserted AP
15:17:17.704 -> - SRAM left: 5431
15:17:17.704 -> inserting AP:8
15:17:17.784 -> - SRAM left: 5431
15:17:17.784 -> inserted AP
15:17:17.784 -> - SRAM left: 3375
15:17:17.784 -> inserting AP:9
15:17:17.784 -> - SRAM left: 3375
15:17:17.784 -> inserted AP
15:17:17.784 -> - SRAM left: 3375
15:17:17.909 -> inserting AP:10
15:17:17.909 -> - SRAM left: 3375
15:17:17.909 -> inserted AP
15:17:17.909 -> - SRAM left: 3375
15:17:17.909 -> inserting AP:11
15:17:17.909 -> - SRAM left: 3375
15:17:17.909 -> inserted AP
15:17:18.034 -> - SRAM left: 3375
15:17:18.034 -> inserting AP:12
15:17:18.034 -> - SRAM left: 3375
15:17:18.034 -> inserted AP
15:17:18.034 -> - SRAM left: 3375
15:17:18.034 -> inserting AP:13
15:17:18.034 -> - SRAM left: 3375
15:17:18.204 -> inserted AP
15:17:18.204 -> - SRAM left: 3375
15:17:18.204 -> inserting AP:14
15:17:18.204 -> - SRAM left: 3375
15:17:18.204 -> inserted AP
15:17:18.204 -> - SRAM left: 3375
15:17:18.204 -> inserting AP:15
15:17:18.283 -> - SRAM left: 3375
15:17:18.283 -> inserted AP
15:17:18.283 -> - SRAM left: 3375
15:17:18.283 -> inserting AP:16
15:17:18.283 -> - SRAM left: 3375
// the function has never returned
```

For sure, as you mention, the solution proposed in the PR is not the best, maybe adding a new `scanNetworks` with an `int` argument for limiting the APs in list is better, what do you think?
I, also, tried my sketch with the MKR1010 without problems because, I believe, [the list it's limited to 10](https://github.com/arduino-libraries/WiFiNINA/blob/89ca31626b8fa2526c7a6a95f4214b7092eeaf6c/src/utility/wl_definitions.h#L41).

fabik111 commented:
Hello, I've updated the algorithm and added a couple of optimizations:

- refactor of the `CAccessPoint` members
- remove vectors in the parsing of the list of discovered APs (this saves ~2k of flash)

pennam commented on libraries/WiFiS3/src/WiFi.h:
```diff
@@ -19,24 +19,43 @@
 
 #define WIFI_FIRMWARE_LATEST_VERSION "0.4.1"
 
+#define WL_MAX_AP_LIST 10
```
I would move `WL_MAX_AP_LIST` to WiFi.cpp file and rename it to something like `WIFI_MAX_SSID_COUNT` adding the possibility to override it

```
#ifndef WIFI_MAX_SSID_COUNT
  #define WIFI_MAX_SSID_COUNT 10
#endif
```

pennam commented on libraries/WiFiS3/src/WiFi.cpp:
```diff
@@ -264,30 +300,56 @@ int8_t CWifi::scanNetworks() {
-      for(uint16_t i = 0; i < aps.size(); i++) {
+      char *startAp = (char*)res.c_str();
+      char *endAP = strstr(startAp, "\r\n");
+      for(; endAP != NULL; startAp = endAP, endAP = strstr(startAp, "\r\n")) {
```
I found this loop not easy to read and I've tried to rewrite it

```
      for(; endAP != NULL; startAp = endAP, endAP = strstr(startAp, "\r\n")) {
         /* split the modem response in multiple lines and parse once at time.
          * The output will be something like:
          * SSID | BSSID | RSSI | CHANNEL | SECURITY
          */
         *endAP = '\0'; // Replace \r with \0
         endAP++;       // Skip \n

         char *token[5];
         uint8_t i = 1;
         token[0] = startAp;
         for(; i < 5; i++){
            char *endToken = strstr(token[i-1], " | ");
            if(endToken == NULL){
               break;
            }
            memcpy(endToken, "\0", 3);
            token[i] = endToken + 3;
         }

         if(i < 5 || strlen(token[0]) == 0 || strlen(token[0]) > WL_SSID_MAX_LENGTH ||
            strlen(token[1]) != WL_MAX_BSSID_LENGTH ||
            strlen(token[2]) == 0 || strlen(token[3]) == 0 || strlen(token[4]) == 0){
            /* Skip the row and process the next one */
            continue;
         }

         CAccessPoint ap;
         strcpy(ap.ssid, token[0]);
         macStr2macArray(ap.uint_bssid, token[1]);
         ap.rssi = atoi(token[2]);
         ap.channel = atoi(token[3]);
         ap.encryption_mode = Encr2wl_enc(token[4]);

         // insert in list
         if( _apsFound < WL_MAX_AP_LIST ){
            access_points[_apsFound] = ap;
            _apsFound++;
            _sortAPlist(_apsFound);
         }else{
            if (ap.rssi > access_points[WL_MAX_AP_LIST-1].rssi){
               access_points[WL_MAX_AP_LIST-1] = ap;
               _sortAPlist(WL_MAX_AP_LIST);
            }
         }
      }
```

fabik111 replied:
I used a `memset` instead of `memcpy(endToken, "\0", 3);` so all the 3 bytes are setted as '\0'.

fabik111 commented on libraries/WiFiS3/src/WiFi.h:
```diff
@@ -19,24 +19,43 @@
 
 #define WIFI_FIRMWARE_LATEST_VERSION "0.4.1"
 
+#define WL_MAX_AP_LIST 10
```
Can't move to CPP because is needed as size for the access_points array but renamed according your suggestion

fabik111 commented on libraries/WiFiS3/src/WiFi.h:
```diff
@@ -19,24 +19,43 @@
 #define WIFI_FIRMWARE_LATEST_VERSION "0.4.1"
 
+#define WL_MAX_AP_LIST 10
+#define WL_MAX_BSSID_LENGTH 17
```
Implemented

fabik111 commented on libraries/WiFiS3/src/WiFi.h:
```diff
@@ -19,24 +19,43 @@
 
+#define WL_MAX_AP_LIST 10
+#define WL_MAX_BSSID_LENGTH 17
+#define WL_MAX_SSID_LENGHT_S WL_SSID_MAX_LENGTH + 1 // +1 for null terminator
```
Implemented

fabik111 commented on libraries/WiFiS3/src/WiFi.cpp:
```diff
@@ -264,30 +300,56 @@ int8_t CWifi::scanNetworks() {
-            ap.encryption_mode = tokens[4];
-            access_points.push_back(ap);
+         char *token[5];
+         *endAP++ = '\0'; // Replace \r with \0
```
I refactorized moving the incrementation inside the for statement updating section.

fabik111 commented on libraries/WiFiS3/src/WiFi.cpp:
```diff
@@ -264,30 +300,56 @@ int8_t CWifi::scanNetworks() {
+         }
+
+         if(i>=5) {
+            if(strlen(token[0]) > WL_SSID_MAX_LENGTH || strlen(token[1]) != WL_MAX_BSSID_LENGTH){
```
Done
